cpp/src/arrow/error_or.h

```diff
@@ -0,0 +1,440 @@
+  ErrorOr(U&& value)  // NOLINT(runtime/explicit)
+      : variant_(std::forward<U>(value)), has_value_(true) {}
+
+  /// Copy constructor.
+  ///
+  /// This constructor needs to be explicitly defined because the presence of
+  /// the move-assignment operator deletes the default copy constructor. In such
+  /// a scenario, since the deleted copy constructor has stricter binding rules
+  /// than the templated copy constructor, the templated constructor cannot act
+  /// as a copy constructor, and any attempt to copy-construct a `ErrorOr`
+  /// object results in a compilation error.
+  ///
+  /// \param other The value to copy from.
+  ErrorOr(const ErrorOr& other)
+      : has_value_(other.has_value_) {  // NOLINT(runtime/explicit)
+    if (has_value_) {
+      // Re-use the memory from variant when constructing
+      new (&variant_) variant(other.variant_.value_);
+    } else {
+      new (&variant_) variant(other.variant_.status_);
+    }
+  }
+
+  /// Templatized constructor that constructs a `ErrorOr<T>` from a const
+  /// reference to a `ErrorOr<U>`.
+  ///
+  /// `T` must be implicitly constructible from `const U &`.
+  ///
+  /// \param other The value to copy from.
+  template <typename U,
+            typename E = typename std::enable_if<std::is_constructible<T, U>::value &&
+                                                 std::is_convertible<U, T>::value>::type>
+  ErrorOr(const ErrorOr<U>& other) : has_value_(other.has_value_) {
+    if (has_value_) {
+      // Re-use the memory from variant when constructing
+      new (&variant_) variant(other.variant_.value_);
+    } else {
+      new (&variant_) variant(other.variant_.status_);
+    }
+  }
+
+  /// Copy-assignment operator.
+  ///
+  /// \param other The ErrorOr object to copy.
+  ErrorOr& operator=(const ErrorOr& other) {
+    // Check for self-assignment.
+    if (this == &other) {
+      return *this;
+    }
+
+    // Construct the variant object using the variant object of the source.
+    if (other.has_value_) {
+      AssignValue(other.variant_.value_);
+    } else {
+      AssignStatus(other.variant_.status_);
+    }
+    return *this;
+  }
+
+  /// Templatized constructor which constructs a `ErrorOr<T>` by moving the
+  /// contents of a `ErrorOr<U>`. `T` must be implicitly constructible from `U
+  /// &&`.
+  ///
+  /// Sets `other` to contain a non-OK status with a`StatusError::Invalid`
+  /// error code.
+  ///
+  /// \param other The ErrorOr object to move from and set to a non-OK status.
+  template <typename U,
+            typename E = typename std::enable_if<std::is_constructible<T, U>::value &&
+                                                 std::is_convertible<U, T>::value>::type>
+  ErrorOr(ErrorOr<U>&& other) : has_value_(other.has_value_) {
+    if (has_value_) {
+      // Re-use the memory from variant when constructing
+      new (&variant_) variant(std::move(other.variant_.value_));
+      other.OverwriteValueWithStatus(
```

Comment:
I don't think this kind of thing is a good idea. A move constructor is supposed to be cheap but this is doing a heap allocation for the non-ok Status' internal state.

Reply:
I agree; a moved-from object should be considered consumed, so giving it an error message like this is an anti pattern. Maybe put this in `#ifndef NDEBUG` as below?

Reply:
Good catch. I did IFNDEF and assigned to not has_value on the else.  This can potentially lead to undefined behavior but thats why we test in debug mode :)

I think the status in asylo must not have the same costs (I had to change this code slightly).

Reply:
With the move to vendored variant, it I added a const char* assignment which should be cheap.